### PR TITLE
Updates to Timestamp signing and verification

### PR DESCRIPTION
**Pull request by Hayden-IO**
* Switch to using the raw signature rather than base64 signature for OCI and blob signing
* For blob signing, write only the timestamp to disk, not the LocalSignedPayload (since that's already written with the bundle. I want to avoid conflating the Rekor bundle with the TS)
* For blob verification, expect only a timestamp in the file. If you don't pass a bundle, you'll need to also pass the signature by flag
* Some nits from the previous PR

Signed-off-by: Hayden Blauzvern <hblauzvern@google.com>

### CODE REVIEW

**Comment by haydentherapper** [comment]
Still working on e2e testing and some more unit tests, but just wanted to get this up early for feedback

**Comment by haydentherapper** [comment]
Now it should be go, would like to add some more tests in verify-blob for completeness, but might do that in a follow up.

File: cmd/cosign/cli/sign/sign_blob.go [hunk 1]
```diff
@@ -67,19 +67,20 @@ func SignBlobCmd(ro *options.RootOptions, ko options.KeyOpts, regOpts options.Re
...
+	var rfc3161Timestamp *cbundle.RFC3161Timestamp
 	if ko.TSAServerURL != "" {
 		clientTSA, err := tsaclient.GetTimestampClient(ko.TSAServerURL)
 		if err != nil {
 			return nil, fmt.Errorf("failed to create TSA client: %w", err)
 		}
-		b64Sig := []byte(base64.StdEncoding.EncodeToString(sig))
 
-		respBytes, err := tsa.GetTimestampedSignature(b64Sig, clientTSA)
+		respBytes, err := tsa.GetTimestampedSignature(sig, clientTSA)
```
Comment (hectorj2f) [inline cmd/cosign/cli/sign/sign_blob.go:77]: Q: why are we using b64sig ? I remember we agreed on using base64 for the blob.

File: cmd/cosign/cli/verify/verify_blob_test.go [hunk 2]
```diff
@@ -552,6 +525,10 @@ func TestVerifyBlob(t *testing.T) {
 				expiredLeafPem, true)},
 			shouldErr: true,
 		},
+		// TODO: Add tests for TSA:
```
Comment (hectorj2f) [inline cmd/cosign/cli/verify/verify_blob_test.go:528]: This basically removes tests that we had in place. I am not sure we want to change functionality and remove tests. How does it work if I only sign/verify a blob without using tlog ?

File: internal/pkg/cosign/tsa/signer.go [hunk 3]
```diff
@@ -58,7 +60,7 @@ func GetTimestampedSignature(sigBytes []byte, tsaClient *tsaclient.TimestampAuth
 		return nil, err
 	}
 
-	fmt.Fprintln(os.Stderr, "Timestamp fetched with time:", ts.Time)
+	fmt.Fprintln(os.Stderr, "Timestamp fetched with time: ", ts.Time)
```
Comment (hectorj2f) [inline internal/pkg/cosign/tsa/signer.go:63]: This is how it looks today: `Timestamp fetched with time: 2022-12-01 17:44:21 +0100 CET` This will add an extra whitespace here.

File: cmd/cosign/cli/verify/verify_blob.go [hunk 4]
```diff
@@ -73,16 +74,16 @@ func (c *VerifyBlobCmd) Exec(ctx context.Context, blobRef string) error {
 	opts := make([]static.Option, 0)
 
 	// Require a certificate/key OR a local bundle file that has the cert.
-	if options.NOf(c.KeyRef, c.CertRef, c.Sk, c.BundlePath, c.RFC3161TimestampPath) == 0 {
-		return fmt.Errorf("please provide a cert to verify against via --certificate or a bundle via --bundle or --rfc3161-timestamp-bundle")
```
Comment (hectorj2f) [inline cmd/cosign/cli/verify/verify_blob.go:77]: This changes the functionality forcing the usage of a bundle path instead of making it optional to passing a rfc3161 timestamp.

File: cmd/cosign/cli/sign/sign_blob.go (hunk 1 above)
Comment (haydentherapper) [reply cmd/cosign/cli/sign/sign_blob.go:77]: We discussed this a bit in https://github.com/sigstore/timestamp-authority/issues/116, but the only reason base64 was used was because the interface exposed the signature as base64. I'd say it's better to use the raw bytes because you don't have any ambiguity (for example, some base64 encodings might allow variable numbers of `=`).

FWIW, every client can make their own decision on what's timestamped as long as the client's verification layer knows what was timestamped (like for example what you implemented for understanding attestations without signatures). But I think Cosign should stick with raw bytes (and this aligns nicely with the future bundle format that doesn't use base64).

cc @znewman01 and @asraa who I've chatted with about this

File: internal/pkg/cosign/tsa/signer.go (hunk 3 above)
Comment (haydentherapper) [reply internal/pkg/cosign/tsa/signer.go:63]: If this is meant to be debug output, I think it's more readable with whitespace, but I'm not super picky if y'all think otherwise.

File: cmd/cosign/cli/verify/verify_blob_test.go (hunk 2 above)
Comment (haydentherapper) [reply cmd/cosign/cli/verify/verify_blob_test.go:528]: I plan to add some more tests in, but the presence of the bundle is just a way to get the signature and/or other timestamp. If you don't use the tlog, then you must provide the cert/sig the same way you currently do, via --certificate/--signature flags.

File: internal/pkg/cosign/tsa/signer.go (hunk 3 above)
Comment (hectorj2f) [reply internal/pkg/cosign/tsa/signer.go:63]: It is fine, I wanted to be sure we are aware that there is a whitespace between the time and the text.

File: cmd/cosign/cli/verify/verify_blob.go (hunk 4 above)
Comment (haydentherapper) [reply cmd/cosign/cli/verify/verify_blob.go:77]: This is intentional. I don't think we should conflate the 3161 timestamp with the on-disk bundle.

To verify an artifact, you need:
* A signature, which can be provided in the "--bundle" or via a "--sig" flag
* A verification key, provided by "--certificate", "--sk", or "--bundle"

I would prefer to not change the on-disk bundle format and treat the timestamp as separate. I think it's a little confusing to bundle the artifact certificate with the timestamp, since those are unrelated. I would prefer that each element be provided separately.

So tl;dr, you would either:
* provide a --certificate flag, --signature flag, and --3161timestamp
* provide a --bundle flag and --3161timestamp

Comment (haydentherapper) [reply cmd/cosign/cli/verify/verify_blob.go:77]: Another reason I wanted to do this is that `--bundle` == `LocalSignedPayload`, and I didn't want to make changes to that struct. To me, it seems only like an optimization to include the 3161 timestamp in `LocalSignedPayload`, and I think simply tacking on a timestamp flag when you want to use the timestamp is cleaner and easier to explain.

Another option would be a different struct for storing the ts locally, like `Local3161Bundle`, but I think this is confusing to add another bundle format especially as we're working on another bundle format for the specs.

File: cmd/cosign/cli/verify/verify_blob_test.go (hunk 2 above)
Comment (haydentherapper) [reply cmd/cosign/cli/verify/verify_blob_test.go:528]: With these changes, it actually requires less tests since this is only an additional flag that affects verification, rather than modifying codepaths that read in the cert/signature.

**Comment by haydentherapper** [comment]
Verified with a local TSA instance:

```
# Sign blob and upload to Rekor
cosign sign-blob --timestamp-server-url http://localhost:3000 --rfc3161-timestamp tsaresponse --tlog-upload=true --output-certificate=blob.cert --output-signature=blob.sig --bundle blob.bundle blob

# Verify with local Rekor bundle
cosign verify-blob --rfc3161-timestamp tsaresponse --bundle blob.bundle --timestamp-cert-chain tsacertchain blob

# Verify with skipping checking Rekor
cosign verify-blob --rfc3161-timestamp tsaresponse --timestamp-cert-chain tsacertchain --certificate blob.cert --signature blob.sig --insecure-skip-tlog-verify --offline blob

# Sign blob without Rekor upload
cosign sign-blob --timestamp-server-url http://localhost:3000 --rfc3161-timestamp tsaresponse --tlog-upload=false --output-certificate=blob.cert --output-signature=blob.sig  blob
```

An unrelated issue is if I skip tlog-upload, then I can't output a certificate (`cosign sign-blob --timestamp-server-url http://localhost:3000 --rfc3161-timestamp tsaresponse --tlog-upload=false --output-certificate=blob.cert --output-signature=blob.sig blob` fails to output a cert), so I can't verify. I'll fix this separately.

File: specs/SIGNATURE_SPEC.md [hunk 5]
```diff
@@ -129,6 +129,11 @@ Gyp4apdU7AXEwysEQIb034aPrTlpmxh90SnTZFs2DHOvCjCPPAmoWfuQUwPhSPRb
 
 For instructions on using the `bundle` for verification, see [USAGE.md](../USAGE.md#verify-a-signature-was-added-to-the-transparency-log).
 
+* `rfc3161timestamp` string
+
+  This OPTIONAL property contains a JSON formatted `RFC3161Timestamp` bundle containing the timestamp response from a
```
Comment (hectorj2f) [inline specs/SIGNATURE_SPEC.md:134]: I suggest to avoid using the term bundle here, as you mentioned it can be confusing.

File: cmd/cosign/cli/verify/verify_blob.go (hunk 4 above)
Comment (hectorj2f) [reply cmd/cosign/cli/verify/verify_blob.go:77]: 👍🏻 It sounds good to me.

**Review by hectorj2f — COMMENTED** [review]
LGTM, I added some nit comments. Ideally, I'd like to see some tests to validate the usage of the rfc3161timestamp with blobs.

File: cmd/cosign/cli/sign/sign_blob.go [hunk 6]
```diff
@@ -98,18 +99,15 @@ func SignBlobCmd(ro *options.RootOptions, ko options.KeyOpts, regOpts options.Re
 		signedPayload.Bundle = cbundle.EntryToBundle(entry)
 	}
 
-	// if bundle is specified, just do that and ignore the rest
-	if ko.RFC3161TimestampPath != "" {
-		signedPayload.Base64Signature = base64.StdEncoding.EncodeToString(sig)
-
-		contents, err := json.Marshal(signedPayload)
+	if ko.RFC3161TimestampPath != "" && rfc3161Timestamp != nil {
```
Comment (znewman01) [inline cmd/cosign/cli/sign/sign_blob.go:102]: Should we handle the case where I ask for the timestamp to be written, but I don't actually get a timestamp?

Feels worth a warning at least.

File: internal/pkg/cosign/tsa/signer.go (hunk 3 above)
Comment (znewman01) [reply internal/pkg/cosign/tsa/signer.go:63]: +1 to adding the whitespace

File: internal/pkg/cosign/tsa/signer.go [hunk 7]
```diff
@@ -84,8 +86,14 @@ func (rs *signerWrapper) Sign(ctx context.Context, payload io.Reader) (oci.Signa
 		return nil, nil, err
 	}
 
-	// Here we get the response from the timestamped authority server
-	responseBytes, err := GetTimestampedSignature([]byte(b64Sig), rs.tsaClient)
+	// create timestamp over raw bytes of signature
+	rawSig, err := base64.StdEncoding.DecodeString(b64Sig)
```
Comment (znewman01) [inline internal/pkg/cosign/tsa/signer.go:90]: Feels a little gross to ask for the base64 signature then immediately decode it. I really want to push *all* encoding to the application boundaries as much as possible.

Can we add a TODO to add a `Signature()` method to `oci.Signature`?

File: pkg/cosign/verify.go [hunk 8]
```diff
@@ -986,52 +986,59 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, co *CheckOpts, rekorCl
 	return true, nil
 }
 
-func VerifyRFC3161Timestamp(ctx context.Context, sig oci.Signature, tsaCerts *x509.CertPool) (bool, error) {
```
Comment (znewman01) [inline pkg/cosign/verify.go:989]: Thank you for this change 🙏  I never figured out what `bundleVerified` means

**Review by znewman01 — COMMENTED** [review]
Some nits but overall great changes

File: cmd/cosign/cli/sign/sign_blob.go (hunk 6 above)
Comment (asraa) [reply cmd/cosign/cli/sign/sign_blob.go:102]: It seems like this drops the TSA timestamp when there's no timestamp path specified? it seems like there should be a default one then -- otherwise I'd think to only set TSAServerURL (and then my timestamp disappears)

File: specs/SIGNATURE_SPEC.md (hunk 5 above)
Comment (asraa) [reply specs/SIGNATURE_SPEC.md:134]: +1 here it's not a bundle anymore of cert/sig as well, so it's just the RFC 3161 response, I think?

File: cmd/cosign/cli/sign/sign_blob.go (hunk 6 above)
Comment (znewman01) [reply cmd/cosign/cli/sign/sign_blob.go:102]: We already do that if you don't specify `--output-signature` or `--output-certificate` 🤷 Maybe a `y/N` prompt would be the right amount of annoying here, you could bypass with `--yes`

I'd also be comfortable just waiting until we cut over to the new bundle format and then making `--bundle` required.

Comment (asraa) [reply cmd/cosign/cli/sign/sign_blob.go:102]: @znewman01 I think the signature is always output to STDOUT though regardless, so you can always pass it.

I guess it's fair that the certificate is swallowed when you don't specify a output-certificate or bundle even if you use keyless, that seems error-prone to me too.

Comment (asraa) [reply cmd/cosign/cli/sign/sign_blob.go:102]: (But also the cert is written to STDERR at some point during the command output)

File: cmd/cosign/cli/sign/sign_blob.go [hunk 9]
```diff
@@ -98,18 +99,15 @@ func SignBlobCmd(ro *options.RootOptions, ko options.KeyOpts, regOpts options.Re
...
+		ts, err := json.Marshal(rfc3161Timestamp)
 		if err != nil {
 			return nil, err
 		}
-		if err := os.WriteFile(ko.RFC3161TimestampPath, contents, 0600); err != nil {
+		if err := os.WriteFile(ko.RFC3161TimestampPath, ts, 0600); err != nil {
```
Comment (haydentherapper) [inline cmd/cosign/cli/sign/sign_blob.go:107]: Not to self, should require that both TSAServerURL and RFC3161TimestampPath is set

File: cmd/cosign/cli/sign/sign_blob.go (hunk 6 above)
Comment (haydentherapper) [reply cmd/cosign/cli/sign/sign_blob.go:102]: I'm going to require that you set both TSAServerURL and RFC3161TimestampPath, to always write the timestamp. If we add in writing to Rekor at some point, we can change the requirement.

File: specs/SIGNATURE_SPEC.md (hunk 5 above)
Comment (haydentherapper) [reply specs/SIGNATURE_SPEC.md:134]: Correct, removed.

**Review by znewman01 — DISMISSED** [review]

File: cmd/cosign/cli/verify/verify_blob.go [hunk 10]
```diff
@@ -73,16 +74,16 @@ func (c *VerifyBlobCmd) Exec(ctx context.Context, blobRef string) error {
...
+	if options.NOf(c.KeyRef, c.CertRef, c.Sk, c.BundlePath) == 0 {
+		return fmt.Errorf("please provide a cert to verify against via --certificate or a bundle via --bundle")
```
[the remark this answers is not in the document]
Comment (haydentherapper) [reply cmd/cosign/cli/verify/verify_blob.go:78]: done

File: cmd/cosign/cli/sign/sign_blob.go [hunk 11]
```diff
@@ -72,19 +72,36 @@ func SignBlobCmd(ro *options.RootOptions, ko options.KeyOpts, payloadPath string
...
+			return nil, fmt.Errorf("rfc3161 timestamp is nil")
+		}
+		ts, err := json.Marshal(rfc3161Timestamp)
+		if err != nil {
+			return nil, err
+		}
+		if err := os.WriteFile(ko.RFC3161TimestampPath, ts, 0600); err != nil {
+			return nil, fmt.Errorf("create rfc3161 timestamp file: %w", err)
+		}
+		fmt.Printf("RFC3161 timestamp bundle written to file %s\n", ko.RFC3161TimestampPath)
```
[the remark this answers is not in the document]
Comment (haydentherapper) [reply cmd/cosign/cli/sign/sign_blob.go:104]: done

File: cmd/cosign/cli/sign/sign_blob.go [hunk 12]
```diff
@@ -72,19 +72,36 @@ func SignBlobCmd(ro *options.RootOptions, ko options.KeyOpts, payloadPath string
...
 		if err != nil {
 			return nil, err
 		}
 
-		signedPayload.RFC3161Timestamp = cbundle.TimestampToRFC3161Timestamp(respBytes)
+		rfc3161Timestamp = cbundle.TimestampToRFC3161Timestamp(respBytes)
+		// TODO: Consider uploading RFC3161 TS to Rekor
+
+		if rfc3161Timestamp == nil {
+			return nil, fmt.Errorf("rfc3161 timestamp is nil")
```
[the remark this answers is not in the document]
Comment (haydentherapper) [reply cmd/cosign/cli/sign/sign_blob.go:95]: This is a just-in-case check, since `TimestampToRFC3161Timestamp` might return nil if `respBytes` is empty for some reason.

**Review by znewman01 — DISMISSED** [review]

**Review by znewman01 — APPROVED** [review]

**Review by hectorj2f — APPROVED** [review]
lgtm
